File: skills/ghostwriter/evals/budget.py
"""Cost cap + mock mode for the ghostwriter LLM eval harness.

Enforces the standing rule for any job that spends LLM money: estimate the spend
up front, hard-cap it in code with a PRE-CALL gate, and provide a `--mock` mode
so the harness logic runs at $0 (and is testable in CI without an API key).

Pure, dependency-free, deterministic — fully unit-testable.
"""
from __future__ import annotations

import os

# Conservative blended USD per 1K tokens (input+output), rounded UP so the cap
# errs on the safe side. Not billing-accurate — a guardrail, not an invoice.
_PRICE_PER_1K = {
    "claude-haiku-4-5": 0.005,
    "claude-sonnet-4-6": 0.018,
    "claude-opus-4-8": 0.090,
}
_DEFAULT_PRICE = 0.02
DEFAULT_MAX_SPEND = 0.50  # USD per run; override with --max-spend


class BudgetExceeded(RuntimeError):
    """Raised by the pre-call gate before a call that would breach the cap."""


def estimate_usd(text: str, model: str, *, output_tokens: int = 600) -> float:
    """Conservative spend estimate for one call.

    ~4 chars/token for the input, plus an assumed output budget. Over-estimates
    on purpose so the gate trips early rather than late.
    """
    in_tokens = max(1, len(text) // 4)
    price = _PRICE_PER_1K.get(model, _DEFAULT_PRICE)
    return round((in_tokens + output_tokens) / 1000 * price, 4)


def mock_enabled(flag: bool) -> bool:
    """Mock if --mock was passed OR no API key is present (fail safe to $0).

    The second clause means CI — which has no ANTHROPIC_API_KEY — never makes a
    live call even if a caller forgets --mock.
    """
    return bool(flag) or not os.environ.get("ANTHROPIC_API_KEY")


class Budget:
    """Cumulative spend tracker with a hard pre-call gate."""

    def __init__(self, max_spend: float = DEFAULT_MAX_SPEND):
        self.max_spend = max_spend
        self.spent = 0.0

    def guard(self, estimate: float) -> None:
        """Abort BEFORE a call whose estimate would push spend over the cap."""
        if self.spent + estimate > self.max_spend:
            raise BudgetExceeded(
                f"Estimated spend ${self.spent + estimate:.4f} would exceed the "
                f"${self.max_spend:.2f} cap. Aborting before the call. "
                f"Raise it with --max-spend if this is intended."
            )

    def record(self, actual: float) -> None:
        """Record actual spend after a successful call."""
        self.spent += actual

File: skills/devlog/skills/devlog/evals/budget.mjs
// Cost control for live eval runs. Ported from ghostwriter's evals/budget.py:
// deliberately over-estimates, guards BEFORE any call that would exceed the
// cap, and forces mock mode when there is no API key so CI can never spend.

export const DEFAULT_MAX_SPEND = 0.50; // USD

// Conservative blended $/1K-token prices by model-family substring. These are
// intentionally pessimistic (treated as if every token cost the output rate).
const PRICE_PER_1K = [
  [/haiku/i, 0.005],
  [/sonnet/i, 0.018],
  [/opus/i, 0.09],
];
const DEFAULT_PRICE_PER_1K = 0.02;

export function pricePer1k(model) {
  for (const [re, price] of PRICE_PER_1K) {
    if (re.test(model || '')) return price;
  }
  return DEFAULT_PRICE_PER_1K;
}

// Over-estimate the cost of one `claude -p` judge call: a context baseline for
// the re-sent system prompt, ~4 chars/token for the prompt itself, plus the
// output budget.
export function estimateUsd(text, model, { outputTokens = 800, contextTokens = 6000 } = {}) {
  const promptTokens = Math.ceil((text || '').length / 4);
  const totalTokens = contextTokens + promptTokens + outputTokens;
  return (totalTokens / 1000) * pricePer1k(model);
}

export class BudgetExceeded extends Error {}

export class Budget {
  constructor(maxSpend = DEFAULT_MAX_SPEND) {
    this.maxSpend = maxSpend;
    this.spent = 0;
  }

  // Call BEFORE spending: throws if the estimate would push cumulative spend
  // over the cap. The cap is a hard ceiling, not advisory.
  guard(estimate) {
    if (this.spent + estimate > this.maxSpend) {
      throw new BudgetExceeded(
        `Refusing call: ~$${estimate.toFixed(3)} would push spend to `
        + `$${(this.spent + estimate).toFixed(3)} > cap $${this.maxSpend.toFixed(2)}`,
      );
    }
  }

  record(actual) {
    this.spent += actual;
  }
}

// Mock is forced when there is no ANTHROPIC_API_KEY — CI has no key, so a
// forgotten --mock flag can never turn into a live spend there.
export function mockEnabled(flag) {
  return Boolean(flag) || !process.env.ANTHROPIC_API_KEY;
}
